backends-velox/src/test/scala/org/apache/gluten/execution/TestOperator.scala

```diff
@@ -1872,4 +1872,16 @@ class TestOperator extends VeloxWholeStageTransformerSuite with AdaptiveSparkPla
         }
     }
   }
+
+  test("Shouldn't pushdown deterministic filter to scan node") {
+    // there are total 60175 rows in lineitem, after rand filter,
+    // there should be around 30000 rows
+    val resultLength = spark
+      .sql("""
+             |select * from lineitem where rand() <= 0.5
+             |""".stripMargin)
+      .collect()
+      .length
+    assert(resultLength >= 25000)
```

Comment:
Could give `resultLength` a upper bound to improve stability of the test? Say if `rand()` yielded a constant `0.1` then the test could still pass as false-positive.

Reply:
It's not velox's rand bug, bug `rand() < 0.5` was evaluate twice,  first is in `ScanNode`, then in `FilterNode`, so that the filtered result is 0.5 * 0.5.

Reply:
> it was because rand() < 0.5 was evaluate twice

In which place it gets evaluated twice? It should be either in subfield filters or in remaining filters already by design. Am I missing something? cc @rui-mo 

Would you mean it's placed both in subfield filters and remaining filters?

Reply:
The design is all filters are pushed down to Spark's Scan operator on Scala side, and in Velox's TableScanNode there are subfield filters and remaining filter to handle two kinds filters according to if their push-down is supported. I'm not clear if the duplication occurs on Scala side between Scan and Filter operator, or on native side between subfield filters and remaining filter. @WangGuangxin Could you clarify?

Reply:
let me double check

Reply:
Take the UT as an example, the `rand() < 0.5` both exists in 
1. `getPushedFilter`  https://github.com/apache/incubator-gluten/blob/9cceba6812d4de9883695f20bc506648083715f9/gluten-core/src/main/scala/org/apache/gluten/backendsapi/SparkPlanExecApi.scala#L672
which is pushed to `ScanNode` ( it has nothing to do with the logic of `remainingFilter` and `subFieldFilter`)

2. and also, the `FilterExecTransformer` also has `rand() < 0.5` in `getRemainingCondition` https://github.com/apache/incubator-gluten/blob/9cceba6812d4de9883695f20bc506648083715f9/backends-velox/src/main/scala/org/apache/gluten/execution/FilterExecTransformer.scala#L39

Reply:
> Take the UT as an example, the `rand() < 0.5` both exists in
> 
> 1. `getPushedFilter`  https://github.com/apache/incubator-gluten/blob/9cceba6812d4de9883695f20bc506648083715f9/gluten-core/src/main/scala/org/apache/gluten/backendsapi/SparkPlanExecApi.scala#L672
>    
>    which is pushed to `ScanNode` ( it has nothing to do with the logic of `remainingFilter` and `subFieldFilter`)
> 2. and also, the `FilterExecTransformer` also has `rand() < 0.5` in `getRemainingCondition` https://github.com/apache/incubator-gluten/blob/9cceba6812d4de9883695f20bc506648083715f9/backends-velox/src/main/scala/org/apache/gluten/execution/FilterExecTransformer.scala#L39

It looks like a bug. By looking into the code, I believe the author didn't want a filter expression appearing in `FilterHandler.getRemainingFilters` once it was pushed into scan. So perhaps we need a bugfix for that.

Reply:
However I will not be against this change if it doesn't bring performance issues. Given that it's a more robust way to make sure non-deterministic filter expressions be only evaluated once from JVM side. 

But I think we still need to find out reason of the bug.

Reply:
> if it doesn't bring performance issues.

But probably it's hardly true in Velox backend if the backend does support pushing non-deterministic exprs to file reader. So my assumption could be too ideal, I am not sure.